File: Recursion/Parenthesis/parenthesis.py
from typing import List

class Solution:
    def doRecursion(self, ans, helper, opening, closed, limit):
        if closed == limit:
            ans.append(helper)
            return
        if opening < limit: self.doRecursion(ans, helper+'(',opening+1,closed,limit)
        if opening>closed: self.doRecursion(ans, helper+')',opening,closed+1,limit)

    def generateParenthesis(self, n: int) -> List[str]:
        ans = []
        self.doRecursion(ans, "",0,0,n)
        return ans

File: Recursion/Parenthesis/parenthesis.cpp
#include <vector>
#include <string>
using namespace std;

class Solution {
public:
    void helper(int n,vector<string> &v,int o = 0,int c = 0,string s = ""){
        if(o==n && c==n){
            v.push_back(s);
            return;
        }
        if(o>c && c<n)helper(n,v,o,c+1,s+")");
        if(o<n)helper(n,v,o+1,c,s+"(");
    }
    vector<string> generateParenthesis(int n) {
        vector<string> s;
        helper(n,s);
        return s;
    }
};